src/catchup/CatchupManagerImpl.cpp

```diff
@@ -93,4 +200,110 @@ CatchupManagerImpl::logAndUpdateCatchupStatus(bool contiguous)
 {
     logAndUpdateCatchupStatus(contiguous, getStatus());
 }
+
+bool
+CatchupManagerImpl::hasBufferedLedger() const
+{
+    return !mSyncingLedgers.empty();
+}
+
+LedgerCloseData const&
+CatchupManagerImpl::getBufferedLedger() const
+{
+    if (!hasBufferedLedger())
+    {
+        throw std::runtime_error(
+            "getBufferedLedger called when mSyncingLedgers is empty!");
+    }
+
+    return mSyncingLedgers.front();
+}
+
+void
+CatchupManagerImpl::popBufferedLedger()
+{
+    if (!hasBufferedLedger())
+    {
+        throw std::runtime_error(
+            "popBufferedLedger called when mSyncingLedgers is empty!");
+    }
+
+    mSyncingLedgers.pop_front();
+}
+
+void
+CatchupManagerImpl::syncMetrics()
+{
+    mSyncingLedgersSize.set_count(mSyncingLedgers.size());
+}
+
+void
+CatchupManagerImpl::trimAndReset()
+{
+    assert(mCatchupWork);
+
+    if (mCatchupWork->getState() == BasicWork::State::WORK_SUCCESS)
+    {
+        assert(mSyncingLedgers.empty());
+        logAndUpdateCatchupStatus(true);
```

Comment:
why do we only reset the status on success?

Reply:
My reasoning was that if there was an error, catchup will run again and reset it. The relevant status message is saved until the reset.

We could also just reset the status regardless. I'm fine with either approach.

Reply:
oh I see. yeah that seems fine then (maybe a comment to explain that we want to keep the failure message until it gets restarted)

Reply:
Ended up resetting the status either way to make this simpler.